Actors cannot conform to protocols with parameter pack based requirements

### Description

If I have a protocol with a function that takes a parameter pack of Sendable types, actors cannot conform to that protocol.

### Reproduction


```swift
protocol This {
    func add<each A: Sendable>(_ a: repeat each A) async
}

actor That: This {
    func add<each A: Sendable>(_ a: repeat each A) async {
        for b in repeat (each a) {
            print(b)
        }
    }
}
```
generates this error
```
/var/folders/6v/zhyq15kx1fz1d9hjv1zq9n8c0000gn/T/vscode-swift/Untitled-103afa4b982d71.swift:6:10: error: non-Sendable parameter type 'repeat each A' cannot be sent from caller of protocol requirement 'add' into actor-isolated implementation
 4 | 
 5 | actor That: This {
 6 |     func add<each A: Sendable>(_ a: repeat each A) async {
   |          `- error: non-Sendable parameter type 'repeat each A' cannot be sent from caller of protocol requirement 'add' into actor-isolated implementation
 7 |         for b in repeat (each a) {
 8 |             print(b)
```

### Expected behavior

I would expect to be able to conform an actor to the protocol and that `repeat each A` is Sendable because each element is Sendable

### Environment

Apple Swift version 6.2-dev (LLVM 5fbc818cf26c90b, Swift 2e1897356956e43)


### Additional information

_No response_

--- a/lib/Sema/TypeCheckConcurrency.cpp
+++ b/lib/Sema/TypeCheckConcurrency.cpp
@@ -1052,10 +1052,16 @@ bool swift::diagnoseNonSendableTypes(
     Type type, SendableCheckContext fromContext,
     Type inDerivedConformance, SourceLoc loc,
     llvm::function_ref<bool(Type, DiagnosticBehavior)> diagnose) {
+  auto &ctx = type->getASTContext();
+
   // If the Sendable protocol is missing, do nothing.
-  auto proto = type->getASTContext().getProtocol(KnownProtocolKind::Sendable);
+  auto proto = ctx.getProtocol(KnownProtocolKind::Sendable);
   if (!proto)
     return false;
+
+  // Unwrap pack expansions here to allow packs of Sendable type.
+  if (auto *expansion = type->getAs<PackExpansionType>())
+    type = PackType::get(ctx, {expansion});
 
   // FIXME: More detail for unavailable conformances.
   auto conformance = lookupConformance(type, proto, /*allowMissing=*/true);
@@ -3054,12 +3060,6 @@ namespace {
         Type type = getDeclContext()
             ->mapTypeIntoContext(decl->getInterfaceType())
             ->getReferenceStorageReferent();
-
-        // Pack expansions are okay to capture as long as the pattern
-        // type is Sendable.
-        if (auto *expansion = type->getAs<PackExpansionType>()) {
-          type = expansion->getPatternType();
-        }
 
         if (type->hasError())
           continue;

--- a/test/Concurrency/sendable_checking.swift
+++ b/test/Concurrency/sendable_checking.swift
@@ -534,3 +534,27 @@ func test(value: (_: Int, _: () -> Void)) {
   takesSendable(value.1) // Ok
   // expected-warning@-1 {{converting non-Sendable function value to '@Sendable () -> Void' may introduce data races}}
 }
+
+// Don't forget about parameter packs -- https://github.com/swiftlang/swift/issues/82614
+
+@available(SwiftStdlib 5.1, *)
+protocol PackProto {
+  func foo<each A: Sendable>(_ a: repeat each A) async
+  func bar<each A>(_ a: repeat each A) async
+  // expected-note@-1 {{consider making generic parameter 'each A' conform to the 'Sendable' protocol}}
+}
+
+@available(SwiftStdlib 5.1, *)
+actor PackActor: PackProto {
+  func foo<each A: Sendable>(_ a: repeat each A) async {
+    for b in repeat (each a) {
+      print(b)
+    }
+  }
+  func bar<each A>(_ a: repeat each A) async {
+  // expected-warning@-1 {{non-Sendable parameter type 'repeat each A' cannot be sent from caller of protocol requirement 'bar' into actor-isolated implementation; this is an error in the Swift 6 language mode}}
+    for b in repeat (each a) {
+      print(b)
+    }
+  }
+}

